### PR TITLE
Online translation

### PR DESCRIPTION
**Pull request by litvinovg**
JIRA Issue [VIVO-2002](https://jira.lyrasis.org/browse/VIVO-2002)
# What does this pull request do?
Adds option for VIVO and Vitro localization on website with downloading results in .properties file. This should simplify process of Vitro and VIVO interface translation.

# What's new?
Added option to export translations with keys and arguments to HTML and parse HTML with javascripts to provide and interface for convinient translation of *.properties files in website context.

# How should this be tested?
Enable online translation checkbox in Developer's panel
Try to translate VIVO or Vitro instance, download saved results (Export all), update properties files from Vitro and VIVO (Update file), import saved results from saved file and clean translations.

# Interested parties
@VIVO-project/vivo-committers

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 	private static final Log log = LogFactory.getLog(I18nBundle.class);
 	private static final String startSep = "\u25a4";
 	private static final String endSep = "\u25a5";
-	private static final String intSep = "\u25a6";
+	public static final String intSep = "\u25a6";
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java:28]: Personally, I prefer naming constants in accordance with Java naming conventions - https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html. Can you please apply that convention here? For instance, START_SEP.

File: webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl [hunk 2]
```diff
@@ -19,6 +19,9 @@
 <#elseif !settings.mayControl>
 	<div class="developer">
 		<h1>${siteName} is running in developer mode.</h1>
+		<div id="publicSettings" style="display:none;">
+			<@showCheckbox "developer_i18n_onlineTranslation","Enable online translation" />
+		</div>
```
Comment (chenejac) [inline webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl:24]: Apologies, I didn't get the purpose of this code. Can you please check if this code is necessary?

**Review by chenejac — CHANGES_REQUESTED** [review]
It works as described. However, I have added a couple of small comments. @litvinovg can you please check those comments before we merge this PR into the repository?

File: webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl (hunk 2 above)
Comment (litvinovg) [reply webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl:24]: It is an option in developer's panel to turn on/off online translations

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java (hunk 1 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java:28]: Sure

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/freemarker/I18nStringTemplateModel.java [hunk 3]
```diff
@@ -75,5 +83,29 @@ public Object exec(List args) throws TemplateModelException {
...
+	 * @param arguments that should be listed before message and substituted in the message itself 
+	 * @return
+	 */
+	private String getOnlineTranslationsFormattedMessage(String preProcessed, Object[] unwrappedArgs) {
+		String[] parts = preProcessed.split(I18nBundle.intSep);
+		final int messageIndex = parts.length -1;
+		String message = MessageFormat.format(parts[messageIndex], unwrappedArgs);
+		String[] arguments = Arrays.copyOf(unwrappedArgs, unwrappedArgs.length, String[].class);
+		parts[messageIndex] = "";
+		String result = String.join(I18nBundle.intSep, parts) +
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/freemarker/I18nStringTemplateModel.java:100]: It is not the same.  String.join(I18nBundle.intSep, parts) doesn't have last element of preProcessed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java (hunk 1 above)
Comment (chenejac) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/I18nBundle.java]: Thanks

File: webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl (hunk 2 above)
Comment (chenejac) [reply webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl:24]: And what is the purpose of lines 62 and 63 of this file?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/freemarker/I18nStringTemplateModel.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chenejac) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/i18n/freemarker/I18nStringTemplateModel.java]: ok, apologies

File: webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl (hunk 2 above)
Comment (litvinovg) [reply webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl:24]: Oh. Sorry I got mixed up. I forgot about it as it was a long time ago. This code is here to make online translations work for unauthorized users. When user is not authorized there is no developer panel on page, so part of the online translation code that runs in browser won't know if it's on or off.  And the 62-63 lines is to turn it on and off when the dev panel is on page.

Comment (chenejac) [reply webapp/src/main/webapp/templates/freemarker/page/partials/developerPanel.ftl:24]: Thanks for clarification.

**Review by chenejac — APPROVED** [review]

**Review by wwelling — APPROVED** [review]
There are some spurious whitespaces. Additional concerns of maintaining JavaScript of this nature without a more modern JavaScript dependency management and bundle process.  Otherwise, looks good to me.

**Pull request merged**